### PR TITLE
port window functions

### PR DESCRIPTION
**Pull request by brotzeit**

### CODE REVIEW

File: rust_src/src/lisp.rs [hunk 1]
```diff
@@ -899,6 +899,17 @@ impl From<LispWindowRef> for LispObject {
     }
 }
 
+impl From<LispObject> for Option<LispWindowRef> {
+    #[inline]
+    fn from(o: LispObject) -> Self {
```
Comment (shaleh) [inline rust_src/src/lisp.rs]: We pay for this check twice. Why not use as_window?

Comment (brotzeit) [reply rust_src/src/lisp.rs]: True. Maybe we should start introducing result types now ?

**Comment by brotzeit** [comment]
@db48x is already working on this and we have several similiar functions that also need updating when we have a solution. This PR is fixing a bug so I think it would be good if we would merge it before applying the changes.

File: rust_src/src/windows.rs [hunk 2]
```diff
@@ -622,4 +622,96 @@ pub extern "C" fn CURRENT_MODE_LINE_HEIGHT(mut window: LispWindowRef) -> i32 {
...
+/// MINIBUF t means include the minibuffer window, even if it isn't active.
+/// MINIBUF nil or omitted means include the minibuffer window only
+/// if it's active.
+/// MINIBUF neither nil nor t means never include the minibuffer window.
+#[lisp_fn(min = "0")]
+pub fn window_list(
+    frame: Option<LispFrameRef>,
+    minibuf: LispObject,
+    window: Option<LispWindowRef>,
+) -> LispObject {
```
Comment (shaleh) [inline rust_src/src/windows.rs]: If you wrote the following `if` as a `match` you can avoid the `is_none` and the `unwrap`.

File: rust_src/src/windows.rs [hunk 3]
```diff
@@ -622,4 +622,96 @@ pub extern "C" fn CURRENT_MODE_LINE_HEIGHT(mut window: LispWindowRef) -> i32 {
...
+        if let Some(f) = frame {
+            f.selected_window()
+        } else {
+            selected_window()
+        }
+    } else {
+        window.unwrap().as_lisp_obj()
+    };
+
+    let f_obj: LispObject;
```
Comment (shaleh) [inline rust_src/src/windows.rs]: Why is this not assigned using the if/else?

File: rust_src/src/windows.rs [hunk 4]
```diff
@@ -622,4 +622,96 @@ pub extern "C" fn CURRENT_MODE_LINE_HEIGHT(mut window: LispWindowRef) -> i32 {
...
+
+    unsafe {
+        LispObject::from_raw(window_list_1(
+            w_obj.to_raw(),
+            minibuf.to_raw(),
+            f_obj.to_raw(),
+        ))
+    }
+}
+
```
Comment (shaleh) [inline rust_src/src/windows.rs]: What is with this function? You are not using it.

Comment (brotzeit) [reply rust_src/src/windows.rs]: `window_list_one` ? It's lisp function ?

Comment (shaleh) [reply rust_src/src/windows.rs]: I missed the name in the `lisp_fn` macro. Never mind.

**Comment by brotzeit** [comment]
@Wilfred we already have similiar impls in lisp.rs for other types. I think @db48x is already working on a solution using result types.

IMO we should merge this PR first and then fix all occurrences of this issue at once.

**Comment by shaleh** [comment]
We don't know when the Result experiment will land. If it does. Let's get this one right until then.

**Pull request merged**